### PR TITLE
Refactors build_features to improve performance

### PR DESCRIPTION
**Pull request by cp2boston**
### Pull Request Description
This PR contains refactors within the `build_features` code path that are focused on improving the performance.  There are some new files:
- Dockerfile.arm is included for reference.  It was used to create a testing/debugging environment on a Mac M1
- .dockerignore filters out unwanted context
- tests/profiling/dfs_profile.py is a helper module to run cProfile against DFS

### CODE REVIEW

File: Dockerfile.arm
Comment (gsheni) [inline Dockerfile.arm:1]: Can we put this in the docs (https://featuretools.alteryx.com/en/stable/install.html) or somewhere else? If possible, I would like to keep the root directory clean and minimal.

File: featuretools/tests/profiling/dfs_profile.py
Comment (sbadithe) [inline featuretools/tests/profiling/dfs_profile.py:18]: Just out of curiosity: when is it better to use `import featuretools as ft ` v.s. `from featuretools import ... `? The tests in `featuretools/tests` mostly use `from ... import ...`. Is there any one in particular that should be favored going forward?

File: docs/source/release_notes.rst
Comment (sbadithe) [inline docs/source/release_notes.rst:19]:
```suggestion
    :user:`cp2boston`, :user:`gsheni`
```

File: featuretools/tests/profiling/dfs_profile.py
Comment (cp2boston) [reply featuretools/tests/profiling/dfs_profile.py:18]: tbh, I was just following the pattern in the example docs.  There's no particular benefit other than keeping things in separate namespaces

File: Dockerfile.arm
Comment (cp2boston) [reply Dockerfile.arm:1]: How about a `docker` or `dockerfiles` subdirectory?  Or, I can remove it and update gitignore to not include it

File: docs/source/release_notes.rst
Comment (cp2boston) [reply docs/source/release_notes.rst:19]: +1

File: Dockerfile.arm
Comment (cp2boston) [reply Dockerfile.arm]: I removed it and will keep it locally for my own use

Comment (cp2boston) [reply Dockerfile.arm]: I removed it and will keep it locally for my own use

Comment (gsheni) [reply Dockerfile.arm]: That works.

Another option is to put it in [contributing.md](https://github.com/alteryx/featuretools/blob/main/contributing.md).

File: docs/source/release_notes.rst
Comment (thehomebrewnerd) [inline docs/source/release_notes.rst:12]: This should probably be in the testing changes section.

File: featuretools/feature_base/feature_base.py
Comment (thehomebrewnerd) [inline featuretools/feature_base/feature_base.py:22]: Remove?

File: featuretools/feature_base/feature_base.py
Comment (thehomebrewnerd) [inline featuretools/feature_base/feature_base.py:153]: A walrus in the wild - don't see many of those in our codebase!

File: featuretools/feature_base/feature_base.py
Comment (thehomebrewnerd) [inline featuretools/feature_base/feature_base.py:156]: I think you could also do `ignored = ignored or set()` here if you wanted.

Comment (cp2boston) [reply featuretools/feature_base/feature_base.py:156]: Good call!

File: featuretools/feature_base/feature_base.py
Comment (cp2boston) [reply featuretools/feature_base/feature_base.py:153]: ikr?  If cleans things up rather nicely, i think

File: docs/source/release_notes.rst
Comment (cp2boston) [reply docs/source/release_notes.rst:12]: will do

File: featuretools/feature_base/feature_base.py
Comment (cp2boston) [reply featuretools/feature_base/feature_base.py:22]: Good catch!

**Comment by gsheni** [comment]
@cp2boston The tests mostly pass now but some lines are not being hit. Also there is a conflict

File: featuretools/synthesis/deep_feature_synthesis.py
Comment (rwedge) [inline featuretools/synthesis/deep_feature_synthesis.py:998]: `transform_stacking` as parameter name is confusing.  Currnetly, if `transform_stacking` is `True`, features that stack on top of transform features are not allowed, and when `transform_stacking` is `False`, stacking on transform feature is allowed.

Also, sinces `check_transform_stacking` excludes any combination of inputs where any input is a transform feature or a direct feature of a transform feature, I think we can filter out those two groups of features as valid input types, period, before creating input combinations (either in `_features_by_type` or by making `check_transform_stacking` a `feature_filter` for `_get_matching_inputs` when called from `_build_transform_features`

Comment (cp2boston) [reply featuretools/synthesis/deep_feature_synthesis.py:998]: I'm going to take the feature_filter approach and see how that plays out

File: featuretools/synthesis/deep_feature_synthesis.py
Comment (rwedge) [inline featuretools/synthesis/deep_feature_synthesis.py:844]: These lines should probably stay, it's this removal that is causing the discrepancy in number of features generated

File: featuretools/feature_base/feature_base.py
Comment (ozzieD) [reply featuretools/feature_base/feature_base.py]: Never knew about these till now. I love it!

File: featuretools/synthesis/deep_feature_synthesis.py
Comment (sbadithe) [reply featuretools/synthesis/deep_feature_synthesis.py]: I might have confused myself here but I think the two expressions are logically equivalent and therefore should always stop at the same point.

Comment (cp2boston) [reply featuretools/synthesis/deep_feature_synthesis.py]: any() will stop at the first match, all() has to run through all of them.  Performance-wise, it makes a difference

File: docs/source/release_notes.rst
Comment (cp2boston) [reply docs/source/release_notes.rst:23]: fixed

File: featuretools/synthesis/deep_feature_synthesis.py
Comment (cp2boston) [reply featuretools/synthesis/deep_feature_synthesis.py:844]: Ah, good catch!

File: featuretools/synthesis/deep_feature_synthesis.py
Comment (sbadithe) [reply featuretools/synthesis/deep_feature_synthesis.py]: Oops, I was under the impression `all` stopped at the first `False`

File: docs/source/release_notes.rst
Comment (cp2boston) [reply docs/source/release_notes.rst:10]: I think merge from main got confused.  Added it back in

File: docs/source/release_notes.rst
Comment (cp2boston) [reply docs/source/release_notes.rst:23]: Fixed

File: featuretools/synthesis/deep_feature_synthesis.py
Comment (cp2boston) [reply featuretools/synthesis/deep_feature_synthesis.py]: They are essentially opposites and your assumption is correct.  all() stops at the first false.  For whatever reason, the profile numbers improved when I switched to any()

Comment (sbadithe) [reply featuretools/synthesis/deep_feature_synthesis.py]: Wow, that is fascinating, and not at all what I would have expected! Interesting, thanks!

File: docs/source/release_notes.rst
Comment (thehomebrewnerd) [inline docs/source/release_notes.rst:19]: Correct user name?
image

**Review by thehomebrewnerd — APPROVED** [review]
Seems fine to me, but not sure I completely followed all the refactors. Might be good to get a second opinion before merging.

**Review by rwedge — APPROVED** [review]

**Review by ozzieD — APPROVED** [review]

File: docs/source/release_notes.rst
Comment (gsheni) [inline docs/source/release_notes.rst:19]:
```suggestion
    :user:`cp2boston`, :user:`gsheni`
```

File: docs/source/release_notes.rst
Comment (gsheni) [inline docs/source/release_notes.rst:9]:
```suggestion
        * Refactor build_features to speed up long running DFS calls by 50% (:pr:`2224`)
```

**Pull request merged**
